packages/eslint-plugin/src/rules/sort-interface-members.ts

```diff
@@ -0,0 +1,53 @@
+/**
+ * @fileoverview Forbids unsorted interface members
+ */
+
+import { TSESTree } from '../../../typescript-estree/dist/parser';
+import * as util from '../util';
+import {
+  Identifier,
+  TSPropertySignature
+} from '@typescript-eslint/typescript-estree/dist/ts-estree/ts-estree';
+
+type Options = [];
+type MessageIds = 'notSorted';
+
+export default util.createRule<Options, MessageIds>({
+  name: 'sort-interface-members',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Forbids unsorted interface members',
+      category: 'Stylistic Issues',
+      recommended: false
+    },
+    schema: [],
+    messages: {
+      notSorted: 'The interface members are not sorted alphabetically.'
+    }
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      TSInterfaceBody(node: TSESTree.TSInterfaceBody) {
+        const members = node.body as TSPropertySignature[];
```

Comment:
This cast isn't valid - the body is of type `TypeNode[]` because all 5 of those types are valid interface members.

Your code will crash on the latter 3 cases below because `members[i].key === undefined`
[repl in ast explorer](https://astexplorer.net/#/gist/2c8506d83eea1d350bcef96057ab1e58/7659365f519a4f366666f92e025c885352de35eb)
```TS
interface Foo {
  a : b;
  foo() : void;
  [a: string]: number;
  new () : Bar;
  () : Baz;
}
```

Reply:
Good point! I'll fix that

Reply:
I see 2 possible sorting algorithms:

## Example input:
```
() : Baz;
[a: string]: number;
a : b;
foo() : void;
new () : Bar;
```

We could try to sort the members using the following names:
```
// if TSPropertySignature or TSMethodSignature
member.key.name // a or foo

// TSIndexSignature
member.parameters.name // a

// if TSConstructSignatureDeclaration or TSCallSignatureDeclaration
member.returnType.typeAnnotation.typeName.name // Bar or Baz
```

I see 2 basic approaches on how to sort these:

## Sorting algorithm 1 (ignoring the type)
### Result: 
```
a : b;
[a: string]: number;
new () : Bar;
() : Baz;
foo() : void;
```

### Alternative result:
```
[a: string]: number;
a : b;
new () : Bar;
() : Baz;
foo() : void;
```

In this algorithm, we would need to find a way to sort members with the same name but a different type.

## Sorting algorithm 2 (sorting by type, then alphabetically):

If the order were `TSPropertySignature > TSMethodSignature > isIndexSignature > TSConstructSignatureDeclaration > TSCallSignatureDeclaration`, the result would look as follows.

### Result:
```
a : b;
foo() : void;
[a: string]: number;
new () : Bar;
() : Baz;
```

I guess in the end, we would have to provide the same configurability `[member-ordering](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/member-ordering.md) comes with. Might be another point in favor of extending `member-ordering` instead of creating a new rule.